logstash-core/lib/logstash/pipeline.rb

```diff
@@ -296,9 +339,30 @@ def shutdown(&before_stop)
 
     before_stop.call if block_given?
 
+    @logger.info "Closing inputs"
     @inputs.each(&:do_stop)
+    @logger.info "Closed inputs"
   end # def shutdown
 
+  # After `shutdown` is called from an external thread this is called from the main thread to
+  # tell the worker threads to stop and then block until they've fully stopped
+  # This also stops all filter and output plugins
+  def shutdown_workers
+    # Each worker thread will receive this exactly once!
+    @worker_threads.each do |t|
+      @logger.debug("Pushing shutdown", :thread => t)
+      @input_queue.push(LogStash::SHUTDOWN)
+    end
+
+    @worker_threads.each do |t|
+      @logger.debug("Shutdown waiting for worker thread #{t}")
+      t.join
+    end
+
+    @filters.each(&:do_close)
```

Comment:
I agree that it should go in a future PR.

Thing is, we'll probably have to wrap and re-raise in the best case. If someone tries to shut down a pipeline and it raises an exception, they're kinda screwed because its now in a permanent bad state.